### PR TITLE
Feature/restructure input validation 420

### PR DESCRIPTION
**Pull request by pc532627**
### PR Type

- Feature

### Description

Input validation has been adjusted to adhere to new requirements. Note that several inputs are not validated at all now, and users could get errors that we do not control when using unexpected inputs.

### How Has This Been Tested?

All unit tests run and adjusted as needed.

### Does this PR introduce a breaking change?

N/A

### Checklist before requesting a review
- [x] I have made sure that my PR is not a duplicate.
- [x] My code follows the style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have performed a self-review of my code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by pc532627** [comment]
@gchqdev227 - Code updated inline with ideal input validation method, can you take a look when you get a chance?

**Review by gchqdev227 — CHANGES_REQUESTED** [review]
A couple *incredibly minor* gripes, but this is really good stuff!

File: coreax/approximation.py
Comment (pc532627) [reply coreax/approximation.py:206]: Removed

File: coreax/approximation.py
Comment (pc532627) [reply coreax/approximation.py:194]: I've removed the `exception` from the raise statements of this form - it has required me to disable some pylint checks, but if this leads to more accurate coverage assessments, I think it's worth it

File: tests/unit/test_approximation.py
Comment (pc532627) [reply tests/unit/test_approximation.py:35]: Thanks!

File: coreax/approximation.py
Comment (pc532627) [reply coreax/approximation.py]: @gchqdev227 I've updated the pyproject.toml with the suggested ignore from coverage, but it does not seem to actually exclude the raise lines. I've tried a few different variants and nothing seems to do the job. Am I missing something silly?

**Comment by pc532627** [comment]
Note - do not merge this branch into main until all input validation and import stye is finished

**Review by gchqdev227 — DISMISSED** [review]
Looks good, thanks for making those changes!

File: coreax/approximation.py
Comment (gchqdev227) [reply coreax/approximation.py]: Ah, I thought it stripped whitespace before checking the regex but apparently not. I've had success with [this](https://github.com/gchq/ConcourseTools/blob/f4d53983a581268c8ca7448082317a22cf5ac2d2/pyproject.toml#L11) in the past, might be worth a go? If it does then please push and I'll re-approve, but if not then suggest we revisit in another branch.

Comment (pc532627) [reply coreax/approximation.py]: That's worked - thanks, it seems the whitespace was the issue, noted for the future.

**Review by gchqdev227 — APPROVED** [review]

**Comment by pc532627** [comment]
This now also closes #451 when merged into main.

**Comment by pc532627** [comment]
A couple of notes for review:

- The unittest workflows that uses the pinned package versions now fail due to JAX changing what type of error is raised in one case. Do we want to update these pinned versions, and should this be part of this PR?
- The import style can now be changed as needed, I expect this should be part of a different PR however as it would make this one quite large to review.

File: tests/unit/test_coresubset.py
Comment (jn177246) [inline tests/unit/test_coresubset.py:1]: Within `test_greedy_body(self)` (line 341) `coreax.validation` is still called causing a unit test to fail.

**Review by jn177246 — CHANGES_REQUESTED** [review]
One occurrence of `coreax.validation` found within codebase. Other than that, it looks good.

File: tests/unit/test_coresubset.py
Comment (pc532627) [reply tests/unit/test_coresubset.py:1]: Thanks for catching this - updated pushed to resolve

**Review by jn177246 — APPROVED** [review]
Looks good.

**Pull request merged**
